static/src/javascripts/projects/commercial/modules/messenger/scroll.js

```diff
@@ -0,0 +1,142 @@
+// @flow
+import { addEventListener } from 'lib/events';
+import detect from 'lib/detect';
+import fastdom from 'lib/fastdom-promise';
+import { register } from 'commercial/modules/messenger';
+
+// An intersection observer will allow us to efficiently send slot
+// coordinates for only those that are in the viewport.
+let w = window;
+let useIO = 'IntersectionObserver' in w;
+let taskQueued = false;
+let iframes = {};
+let iframeCounter = 0;
+let observer;
+let visibleIframeIds;
+
+const reset = (window_: WindowProxy) => {
+    w = window_ || window;
+    useIO = 'IntersectionObserver' in w;
+    taskQueued = false;
+    iframes = {};
+    iframeCounter = 0;
+};
+
+// Instances of classes bound to the current view are not serialised correctly
+// by JSON.stringify. That's ok, we don't care if it's a DOMRect or some other
+// object, as long as the calling view receives the frame coordinates.
+const domRectToRect = rect => ({
+    width: rect.width,
+    height: rect.height,
+    top: rect.top,
+    bottom: rect.bottom,
+    left: rect.left,
+    right: rect.right,
+});
+
+const sendCoordinates = (iframeId, domRect) => {
```

Comment:
It would be great to see all these tiny functions typed. Any chance of doing this?